### ISSUE TITLE
Go: Error when accessing CoreDataOutput of eks.Cluster

### ISSUE THREAD

**Issue by bothra90** [pulumi/pulumi-eks#537]
I added the following to my eks cluster setup project:
```
oidcUrl := eksCluster.Core.OidcProvider().ApplyT(func(oidc *iam.OpenIdConnectProvider) pulumi.StringOutput { return oidc.Url })
ctx.Export("oidcUrl", oidcUrl)
```

where eksCluster is defined as:
```
eksCluster, err := eks.NewCluster(ctx, "eks-cluster", &eks.ClusterArgs{
	VpcId:                  pulumi.String(defaultVpc.Id),
	SubnetIds:              toPulumiStringArray(subnet.Ids),
	InstanceType:           pulumi.String("t2.medium"),
	DesiredCapacity:        pulumi.Int(1),
	MaxSize:                pulumi.Int(1),
	MinSize:                pulumi.Int(1),
	EncryptionConfigKeyArn: key.Arn,
	CreateOidcProvider:     pulumi.Bool(true),
})
```
## Current behavior

I'm getting the following error when running pulumi preview:
```
  pulumi:pulumi:Stack (eks-dev):
    error: program failed: missing one or more required arguments
    exit status 1

    error: an unhandled error occurred: program exited with non-zero exit code: 1
```

When I run the command with `--verbose 6`, the logs show:
```
     pulumi:pulumi:Stack  eks-dev  running..   error: program failed: missing one or more required arguments
I0215 16:57:18.681283   66724 eventsink.go:78] eventSink::Infoerr(<{%reset%}>exit status 1
     pulumi:pulumi:Stack  eks-dev  running..   exit status 1
I0215 16:57:18.700585   66724 source_eval.go:178] EvalSourceIterator ended with an error: an unhandled error occurred: program exited with non-zero exit code: 1
```

## Expected behavior
No error

## Context (Environment)

I'm following the instructions [here](https://aws.amazon.com/blogs/opensource/introducing-fine-grained-iam-roles-service-accounts/) to get IAM roles for my kubernetes services.

## Affected feature
Adding OIDC provider to an EKS cluster and accessing its output value.
I get the same error when trying to accomplish this via another output value:
```
oidcUrl := eksCluster.EksCluster.ApplyT(func(cluster *eks.Cluster) pulumi.StringOutput {
	return cluster.Identities.Index(pulumi.Int(0)).ApplyString(func(c eks.ClusterIdentity) string {
		return *c.Oidcs[0].Issuer
	})
})
ctx.Export("oidcUrl", oidcUrl)
```

**Comment by lblackstone** [comment]
@justinvp Can you take a look at this?

**Comment by bothra90** [comment]
@justinvp: Can you help with this?

**Comment by bothra90** [comment]
@lblackstone, @justinvp : Wondering if you can help with a workaround if not a fix. Thanks!

**Comment by justinvp** [comment]
Hi @bothra90. Sorry for the trouble here. Thanks for opening the issue and your patience! This appears to be an issue in our Go SDK codegen.

All of our `New*` functions check for nil args and return an error when that's the case, e.g.:

https://github.com/pulumi/pulumi-eks/blob/05628f78a5a619e989ece78f44563109fae3cd15/sdk/go/eks/managedNodeGroup.go#L28-L32

However, to support components that return other resources, we need to be able to "rehydrate" those resources, and we do so by calling its associated `New` function, passing the resource's URN as a resource option, and nil for the args, e.g.:

https://github.com/pulumi/pulumi-eks/blob/d343323ac895e5b9529cbfbc6b6f0b3212a682ee/sdk/go/eks/init.go#L27-L28

Always returning an error for nil args is the problem here. We need to update the codegen to not return an error when a URN is passed-in: https://github.com/pulumi/pulumi/blob/22f332e094c5c2cf4afa1a841c3c3b7971a6952c/pkg/codegen/go/gen.go#L1049-L1055. And then re-run the codegen for EKS.

**Comment by bothra90** [comment]
@justinvp : Thanks for getting back. I see various accesses to fields inside `args` in lines 1058-1126 (https://github.com/pulumi/pulumi/blob/22f332e094c5c2cf4afa1a841c3c3b7971a6952c/pkg/codegen/go/gen.go#L1058-L1126). So my guess is that it's not a simple change to just not return an error if a URN is provided, but instead fetch the registered resource using the URN instead of creating a new one.

**Comment by infin8x** [comment]
We will get this fixed during the current milestone (which ends April 2)

**Comment by justinvp** [comment]
The Go SDK codegen bug causing this issue has been fixed. Next we'll need to adopt change in `pulumi-eks` and release a new version containing the fix.

**Comment by justinvp** [comment]
The fix is included in `v0.30.0` of `pulumi-eks`.
